drivers/bicker_ser.c

```diff
@@ -0,0 +1,379 @@
+	}
+
+	upsdebug_hex(3, "bicker_send", buf, buflen);
+	return ret;
+}
+
+static ssize_t bicker_receive_buf(void *dst, size_t dstlen)
+{
+	ssize_t ret;
+
+	ret = ser_get_buf(upsfd, dst, dstlen, 1, 1);
+	if (ret < 0) {
+		upslog_with_errno(LOG_WARNING, "ser_get_buf failed");
+		return ret;
+	} else if ((size_t) ret != dstlen) {
+		upslogx(LOG_WARNING, "ser_get_buf returned %d instead of %d",
+			(int)ret, (int)dstlen);
+		return 0;
+	}
+
+	upsdebug_hex(3, "bicker_receive_buf", dst, dstlen);
+	return ret;
+}
+
+static ssize_t bicker_receive(char cmd, void *dst, size_t dstlen)
+{
+	ssize_t ret;
+	size_t datalen;
+	char buf[BICKER_PACKET];
+
+	ret = bicker_receive_buf(buf, 1);
+	if (ret <= 0) {
+		return ret;
+	} else if (buf[0] != BICKER_SOH) {
+		upslogx(LOG_WARNING, "Received 0x%02X instead of SOH (0x%02X)",
+			(unsigned)buf[0], (unsigned)BICKER_SOH);
+		return 0;
+	}
+
+	ret = bicker_receive_buf(buf + 1, 1);
+	if (ret <= 0) {
+		return ret;
+	}
+
+	datalen = buf[1] - 3;
+
+	ret = bicker_receive_buf(buf + 2, datalen + 3);
+	if (ret <= 0) {
+		return ret;
+	} else if (buf[datalen + 4] != BICKER_EOT) {
+		upslogx(LOG_WARNING, "Received 0x%02X instead of EOT (0x%02X)",
+			(unsigned)buf[datalen + 4], (unsigned)BICKER_EOT);
+		return 0;
+	} else if (buf[3] != cmd) {
+		upslogx(LOG_WARNING, "Commands do not match: sent 0x%02X, received 0x%02X",
+			(unsigned)cmd, (unsigned)buf[1]);
+		return 0;
+	} else if (dstlen < datalen) {
+		upslogx(LOG_ERR, "Not enough space for the payload: %d < %d",
+			(int)dstlen, (int)datalen);
+		return 0;
+	}
+
+	if (dst != NULL) {
+		memcpy(dst, buf + 4, datalen);
+	}
+
+	upsdebug_hex(3, "bicker_receive", buf, datalen + 5);
+	return datalen;
+}
+
+static ssize_t bicker_read_int16(char cmd, int16_t *dst)
+{
+	ssize_t ret;
+
+	ret = bicker_send(cmd, NULL, 0);
+	if (ret <= 0) {
+		return ret;
+	}
+
+	/* XXX: by default, data is stored in little-endian so,
+	 * on big-endian platforms, a byte swap should be needed */
+	return bicker_receive(cmd, dst, 2);
+}
+
+/* For some reason the `seconds` delay (at least on my UPSIC-2403D)
+ * is not honored: the shutdown is always delayed by 2 seconds. This
+ * fixed delay seems to be independent from the state of the UPS (on
+ * line or on battery) and from the dip switches setting.
+ *
+ * As response I get the same command with `0xE0` in the data field.
+ */
+static ssize_t bicker_delayed_shutdown(uint8_t seconds)
+{
+	ssize_t ret;
+	uint8_t response;
+
+	ret = bicker_send('\x32', &seconds, 1);
+	if (ret <= 0) {
+		return ret;
+	}
+
+	ret = bicker_receive('\x32', &response, 1);
+	if (ret > 0) {
+		upslogx(LOG_INFO, "Shutting down in %d seconds: response = 0x%02X",
+			(int)seconds, (unsigned)response);
+	}
+
+	return ret;
+}
+
+static ssize_t bicker_shutdown(void)
+{
+	const char *str;
+	int delay;
+
+	str = dstate_getinfo("ups.delay.shutdown");
+	delay = atoi(str);
+	if (delay > 255) {
+		upslogx(LOG_WARNING, "Shutdown delay too big: %d > 255",
+			delay);
+		delay = 255;
```

Comment:
Just in case, isn't `255` some sort of magical value for the device? I've seen issues with others treating "all bits set" specially (e.g. as not available).

Reply:
The manual just says "min 0 max 255". I used 255 and the UPS behaves as usual, i.e. it ignores the delay and cuts the output after 2 seconds.